typing/typeclass.ml

```diff
@@ -1582,17 +1580,19 @@ let class_infos define_class kind
     begin try
       List.iter2 (Ctype.unify env) cl_params cl_params'
     with Ctype.Unify _ ->
-      raise(Error(cl.pci_loc, env,
-            Bad_parameters (cl_id,
-                            Ctype.newconstr (Path.Pident cl_id)
-                                            cl_params,
-                            Ctype.newconstr (Path.Pident cl_id)
-                                            cl_params')))
+      let dummy_class_type params =
+        let tv = Ctype.newvar () in
+        Ctype.newty(Tobject (tv, ref (Some (Pident ty_id, tv :: params)))) in
+      let ty1 = dummy_class_type cl_params
+      and ty2 = dummy_class_type cl_params' in
+      raise(Error(cl.pci_loc, env, Bad_parameters (ty_id, ty1, ty2)))
     end;
     begin try
       Ctype.unify env ty cl_ty
     with Ctype.Unify _ ->
-      let constr = Ctype.newconstr (Path.Pident cl_id) params in
+      let tv = Ctype.newvar () in
+      let constr =
+        Ctype.newty(Tobject (tv, ref (Some (Pident ty_id, tv :: params)))) in
```

Comment:
And here also.

Reply:
Should we amend this case also?

Reply:
This would be nice, but I don't see a way to modify the `Abbrev_type_clash` error that works well for the present use-case and also for the other use-cases for this error. Maybe you need to define a new error, different from `Abbrev_type_clash`, for this case, that does not require the first `type_expr` parameter (but takes a class name for example?). This could also let you produce a more precise/clear error message.

Reply:
Like https://github.com/ocaml/ocaml/pull/11569/commits/b8cd3dda44289af0b1766c1ce0d4f895363db132? Sorry, I couldn't come up with better error naming and message for this case. Any thought? @garrigue

Reply:
Actually, thinking more about this, I think that my previous advice was wrong: to print a decent explanation for the error message, we really want to print a type looking like `(foo, bar) #c`, and then its expansion. In the case of `Bad_parameters` we had a natural formulation of the error explanation that would only mention `#c`, and not the whole type expression. But here we want the whole type expression, and the only way to do it is to do what you were previously doing in your patch, that is

```
      let constr =
        Ctype.newty(Tobject (tv, ref (Some (Pident ty_id, tv :: params)))) in
```

(question: why are you adding `tv` to the params?)

With the current state of my understanding, reverting to the previous approach that you had implemented seems best.

Reply:
My question is answered by types.mli:

```
  | Tobject of type_expr * (Path.t * type_expr list) option ref
  (** [Tobject (`f1:t1;...;fn: tn', `None')] ==> [< f1: t1; ...; fn: tn >]
      f1, fn are represented as a linked list of types using Tfield and Tnil
      constructors.

      [Tobject (_, `Some (`A.ct', [t1;...;tn]')] ==> [(t1, ..., tn) A.ct].
      where A.ct is the type of some class.

      There are also special cases for so-called "class-types", cf. [Typeclass]
      and [Ctype.set_object_name]:

        [Tobject (Tfield(_,_,...(Tfield(_,_,rv)...),
                         Some(`A.#ct`, [rv;t1;...;tn])]
             ==> [(t1, ..., tn) #A.ct]
        [Tobject (_, Some(`A.#ct`, [Tnil;t1;...;tn])] ==> [(t1, ..., tn) A.ct]

      where [rv] is the hidden row variable.
  *)
```

Reply:
Note: there is a nicer way to define this `constr` type expression:

```ocaml
let constr = Ctype.newobj (Ctype.newvar ()) in
Ctype.set_object_name ty_id params constr;
```

Reply:
Thanks. Reflected at https://github.com/ocaml/ocaml/pull/11569/commits/f8f3ac49137999e38e8a0be5e99fc6600f95fa76